Do not serve stale values from the assets when there is no KV

dario-piotrowicz opened the pull request:
This change makes sure that stale cache values from assets are not served when there is no KV binding.

This makes the issue in https://github.com/opennextjs/opennextjs-cloudflare/issues/328 no longer flaky (it makes the test always fail) since before it was passing because stale cache values were being served from assets (so the fetch cached result would always stay the same, making this [check](https://github.com/opennextjs/opennextjs-cloudflare/blob/420b598ac242ff91768685d312c9940335f3a120/examples/e2e/app-router/e2e/ssr.test.ts#L37) pass). The reason why KV is not being used is still being investigated.

No other tests besides the [328 one](https://github.com/opennextjs/opennextjs-cloudflare/blob/420b598ac242ff91768685d312c9940335f3a120/examples/e2e/app-router/e2e/ssr.test.ts#L31-L38) is affected as far as I know, since the 328 one was the only flaky we had, regarding all the other tests, the ones that were passing before as still passing after my change, and the ones that were failing before (and we skip) are failing after my change.

vicb commented on packages/cloudflare/src/api/kvCache.ts:
```diff
@@ -62,6 +79,17 @@ class Cache implements IncrementalCache {
         }
       }
+
+      const entryValue = entry?.value as CachedFetchValue | undefined;
```
Why do we need `as CachedFetchValue` here? 
Couldn't this be changed upstream?

dario-piotrowicz replied:
I've upstreamed the type, however I still need the `as CachedFetchValue` since typescript is unable to narrow down the type (because of the generic `IsFetch`) 😖

vicb commented on packages/cloudflare/src/api/kvCache.ts:
```diff
@@ -62,6 +79,17 @@ class Cache implements IncrementalCache {
       }
+
+      const entryValue = entry?.value as CachedFetchValue | undefined;
+      if (entryValue?.kind === "FETCH") {
```
Don't we only want to do this when the value comes from the assets?

dario-piotrowicz replied:
why wouldn't we want to check the expiration date also for entries from the KV?

vicb replied:
My understanding is that this is done by the caller.

My understanding of the issue: when an entry is expired, Next will use the [stalled entry](https://nextjs.org/docs/app/building-your-application/caching#time-based-revalidation) and update the cache in the background.

If there is no KV then we can not update the cache and we will keep serving the stalled entry.

Then I think that we don't want to use the stalled entry in this case only.

So I think my previous comment should be

```diff
-Don't we only want to do this when the value comes from the assets?
+Don't we only want to do this when there is no KV?
```

vicb reviewed (commented):
Could you add a comment in the file about what the code is doing an why?

> no longer flaky (it makes the test always fail)

Is that an improvement?

dario-piotrowicz commented:
> Could you add a comment in the file about what the code is doing an why?

ok

> Is that an improvement?

Yes, this is fixing a false positive we currently have, locally things are not working as intended, we are serving state cached data which make the test app-router fetch cache test pass, with this change the state data is no longer served so the underlying broken catching logic gets exposed

dario-piotrowicz commented:
> Is that an improvement?

I thought we discussed this and agreed that not keep always serving the same stale build time cache data was the correct behavior, if you disagree I'm fine closing the PR

vicb commented:
> not keep always serving the same stale build time cache data was the correct behavior, if you disagree I'm fine closing the PR

I agree with that 100%.

My question was if going from flaky to failing is an improvement.

I think the outcome of that PR should be that the tests are passing?

dario-piotrowicz commented on packages/cloudflare/src/api/kvCache.ts:
```diff
@@ -61,7 +66,20 @@ class Cache implements IncrementalCache {
+            this.debug(`found expired entry (expire time: ${expires})`);
+            return null;
+          }
+        }
```
> wondering if
> ...
> Can not be rewritten without the cast as
> ...

No, that still requires some type casting or a type guard

james-elicx replied:
I think a stricter check instead of the type cast would probably be preferable. Something like follows would work:
```ts
if (entryValue && "kind" in entryValue && entryValue.kind === "FETCH") { ... }
```

james-elicx replied:
it would also be nice to get rid of the type cast for `expires as string` too. The existence of the `expires` value could be incorporated into the `if` statement's condition as well.

vicb commented on .changeset/spotty-baboons-rule.md:
```diff
@@ -0,0 +1,5 @@
+---
```
maybe merge the 2 changesets, they are related

dario-piotrowicz replied:
would having the 2 changesets merged have some benefit? or put differently, does having 2 separate changesets have downsides?

vicb replied:
Up to you

vicb commented:
> > My question was if going from flaky to failing is an improvement.
> 
> Well I see it as a temporary improvement, my plan here is to go from flaky-difficult-to-test bug -> consistently failing (easier to work on) bug -> fixed bug
> 
> > I think the outcome of that PR should be that the tests are passing?
> 
> The outcome of this PR is to fix the fact that now we can serve static stale assets forever, as a side effect this makes working on the fetch cache bug easier, that will be addressed in a separate PR (probably #340)

The PR is about not serving stale cache values from assets when there is no KV binding. So I think the title and description of the PR must be updated as well as the changeset.

You could also expand the description of the PR to tell more about why/which of the (skipped) tests always fail after this PR:

- maybe we don't really understand and it's fine to say that we will investigate
- if we have some clue, it would be good to update the linked issue

The kind of information that I am looking for is why this PR would make the tests always fail - it is supposed to change the behavior only when there is no kv binding but we have [a kv binding in the tests](https://github.com/opennextjs/opennextjs-cloudflare/blob/420b598ac242ff91768685d312c9940335f3a120/examples/e2e/app-router/wrangler.json#L11-L16).

I hope this helps.

dario-piotrowicz commented:
> The PR is about not serving stale cache values from assets when there is no KV binding. So I think the title and description of the PR must be updated as well as the changeset.

Updated 👍 

> You could also expand the description of the PR to tell more about why/which of the (skipped) tests always fail after this PR:
> 
> * maybe we don't really understand and it's fine to say that we will investigate
> * if we have some clue, it would be good to update the linked issue
> 
> The kind of information that I am looking for is why this PR would make the tests always fail - it is supposed to change the behavior only when there is no kv binding but we have [a kv binding in the tests](https://github.com/opennextjs/opennextjs-cloudflare/blob/420b598ac242ff91768685d312c9940335f3a120/examples/e2e/app-router/wrangler.json#L11-L16).

I've added the information, I hope it helps

vicb reviewed (approved):
Thanks a lot for the updates, super clear 🙏

dario-piotrowicz commented:
Thanks for the approval @vicb 👍 

We can merge this PR, or if you think it would be safer we can be merged when issue https://github.com/opennextjs/opennextjs-cloudflare/issues/328 is fully understood, either is totally fine by me

vicb commented:
Let's merge it
